clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -104,12 +121,26 @@ bool
 invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
 {
     per_shard_t *shard = reinterpret_cast<per_shard_t *>(shard_data);
-    ++shard->ref_count_;
     if (shard->tid_ == -1 && memref.data.tid != 0)
         shard->tid_ = memref.data.tid;
-    // XXX i#5538: Have the infrastructure provide a common instr and record count.
+    // We check the memtrace_stream_t counts with our own, unless there was an
+    // instr skip from the start where we cannot compare, or we're in a unit
+    // test with no stream interface, or we're in serial mode (since we want
+    // per-shard counts for error reporting; XXX: we could add global local counts).
+    ++shard->ref_count_;
     if (type_is_instr(memref.instr.type))
         ++shard->instr_count_;
+    if (shard->instr_count_ <= 1 && !shard->skipped_instrs_ && shard->stream != nullptr &&
```

Comment:
Even when there's an instr skip later in the trace, the invariant checker won't be able to verify the stream counts, right?

Reply:
Yes.  Not sure how to detect that though.  Added a comment.